### ISSUE TITLE
Request management with Gunicorn sync worker 

### ISSUE THREAD

**Issue by krishnand** [benoitc/gunicorn#2058]
I have an API that reverse proxies requests to a Python backend (both services currently run in the same container). The Python backend is a Flask API app served by Gunicorn with 8 sync workers. My workload has long running APIs that I manage with a task cancelation pattern. However such canceled tasks are still being processed by the Gunicorn workers. How do I efficiently handle this?

1. Is there a recommended task cancelation or timeout pattern for sync workers with Gunicorn.
This is orthogonal to the `--timeout` setting on the worker since this is to only recycle the worker should that worker not accept any request in that time. I see no clean way to do this but I assume the request itself must be time out the request itself.

2.  My other question is what happens when concurrent requests are made to the Gunicorn server? For e.g., if I have 4 sync workers each serving a long running request, what happens to the other requests - are they simply sent to the worker to be handled (in this case by the Flask app) or are these requests still waiting on connection establishment (at the client).?

**Comment by tilgovi** [comment]
1. There is no reliable way to cancel a request once the application starts processing it. HTTP is a request/response protocol. If you want to have long-running tasks with cancellation you should probably make scheduling and canceling tasks explicit routes in your HTTP API and process the tasks using a separate task scheduling system like Celery.

2. If 4 sync workers are handling requests and you have 8 sync workers then 4 will still be available for responding to new requests. If you only have 4 or fewer CPU threads, then the OS will schedule these workers like any other process, and they will share the available CPU.

**Comment by krishnand** [comment]
It is correct that  task cancelation is tricky and a separate sub system to handle / manage tasks is probably the right idea. That said, was there ever a consideration allow for thread aborts with gunicorn especially when using threaded workers (as a server hook maybe ?).

I should have clarified the other question clearly. Assuming 4 sync workers (non-threaded) and 8 concurrent requests - what happens to the other 4 requests that are currently not served - are these still waiting for a connection on the client or they are accepted by gunicorn and sent to the child workers to be handled.

**Comment by tilgovi** [comment]
> It is correct that task cancelation is tricky and a separate sub system to handle / manage tasks is probably the right idea. That said, was there ever a consideration allow for thread aborts with gunicorn especially when using threaded workers (as a server hook maybe ?).

There has not been any plan or proposal for this to date.

> I should have clarified the other question clearly. Assuming 4 sync workers (non-threaded) and 8 concurrent requests - what happens to the other 4 requests that are currently not served - are these still waiting for a connection on the client or they are accepted by gunicorn and sent to the child workers to be handled.

They will still be waiting in the OS listen backlog, waiting for a worker to accept them. Gunicorn does not accept them during this time, only the individual workers do that.
